File: DS LAB/Lab 5/Q1.c
#include<stdio.h>
#include<stdlib.h>
#include<stdbool.h>

#define MAX_SIZE 100

typedef struct STACK{
    char a[MAX_SIZE];
    int n;
    int top;
} stack;

bool isFull(stack* s){
    if(s->top==s->n-1){
        return true;
    }
    else return false;
}

bool isEmpty(stack* s){
    if(s->top==-1){
        return true;
    }
    else return false;
}

void push(stack *s){
    printf("\nEnter element to push: ");
    int element;
    scanf(" %c", &element);
    s->a[++(s->top)]=element;
}

void pop(stack *s){
    printf("\nPopped %c out of stack\n", s->a[s->top]);
    //s->a[s->top]="0";
    s->top--;
}

void display(stack *s){
    printf("\nDisplaying Stack: ");
    for(int i=0; i<=s->top; i++){
        printf("%c ", s->a[i]);
    }
    printf("\n");
}

int main(){
    stack *s=malloc(sizeof(stack));
    s->top=-1;
    printf("Enter size of stack: ");
    scanf("%d", &s->n);

    while(true){
        printf("\n1. Push\n");
        printf("2. Pop\n");
        printf("3. Display\n");
        printf("4. Exit\n");
        printf("Select any option: ");
        int choice;
        scanf("%d", &choice);

        if(choice==1){
            if(!isFull(s)){
                push(s);
            }
            else printf("\nStack Overflow!!\n");
        }

        else if (choice==2){
            if(!isEmpty(s)){
                pop(s);
            }
            else printf("\nStack Underflow!!\n");
        }

        else if(choice==3){
            display(s);
        }
            
        else if (choice==4) break;
            
        else printf("\nSelect valid option\n");
    }
    return 0;
}

File: Sem III/OOP LAB/Lab 8/Q1.java
import java.util.*;

class OverflowException extends Throwable{
    public String toString(){
        return "\nStack overflow!!";
    }
}

class UnderflowException extends Throwable{
    public String toString(){
        return "\nStack underflow!!";
    }
}

class stack{
    int a[];
    int top=-1, size;

    stack(int size){
        this.size = size;
        this.a = new int[size];
    }

    Scanner sc = new Scanner(System.in);

    void push() throws OverflowException{
        if(top==size-1) throw new OverflowException();
        else{
            System.out.print("Enter element: ");
            int element = sc.nextInt();
            a[++top]=element;
        }
    }

    int pop() throws UnderflowException{
        if(top==-1) throw new UnderflowException();
        else{
            int element=a[top--];
            return element;
        }
    }

    void display(){
        System.out.print("Stack: ");
        for(int i=0; i<=top; i++){
            System.out.println(a[i] + " ");
        }
    }
}

class Q1{
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter size of array: ");
        int size=sc.nextInt();
        stack s = new stack(size);
        while(true){            
            System.out.print("\n1. Push \n2. Pop \n3. Display \n4. Exit \nChoice: ");
            int choice = sc.nextInt();

            if(choice==1){
                try{
                    s.push();
                }
                catch(OverflowException e){
                    System.out.println(e);
                }
            }

            else if(choice==2){
                try {
                    System.out.println("Popped: " + s.pop());
                } catch (UnderflowException e) {
                    System.out.println(e);
                }
            }

            else if(choice==3){
                s.display();
            }

            else if(choice==4){
                break;
            }

            else System.out.println("\nEnter a valid option!!\n");
        }
        sc.close();
    }
}
